src/mito2/src/manifest/action.rs

```diff
@@ -196,8 +260,55 @@ impl RegionManifestBuilder {
     }
 }
 
+/// A record of removed files in the region manifest.
+/// This is used to keep track of files that have been removed from the manifest but may still
+/// be kept for a while
+#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq)]
+pub struct RemovedFilesRecord {
+    /// a list of `(FileIds, timestamp)` pairs, where the timestamp is the time when
+    /// the files are removed from manifest. The timestamp is in milliseconds since unix epoch.
+    pub removed_files: Vec<(HashSet<FileId>, i64)>,
+}
+
+impl RemovedFilesRecord {
+    /// Add a record of removed files with the current timestamp.
+    pub fn add_removed_files(&mut self, file_ids: HashSet<FileId>, at: i64) {
+        self.removed_files.push((file_ids, at));
+    }
+
+    pub fn evict_old_removed_files(&mut self, opt: &RemoveFileOptions) -> Result<()> {
+        let total_removed_files: usize = self.removed_files.iter().map(|s| s.0.len()).sum();
+        if total_removed_files <= opt.keep_count {
+            return Ok(());
+        }
+
+        let can_evict_until = chrono::Utc::now()
+            - chrono::Duration::from_std(opt.keep_ttl).context(DurationOutOfRangeSnafu {
+                input: opt.keep_ttl,
+            })?;
+
+        self.removed_files.sort_by_key(|(_, ts)| *ts);
+        let updated = std::mem::take(&mut self.removed_files)
+            .into_iter()
+            .filter_map(|(files, ts)| {
+                if ts < can_evict_until.timestamp_millis() {
+                    // can evict all files
+                    // TODO(discord9): maybe only evict to below keep_count? Maybe not, or the update might be too frequent.
+                    None
+                } else {
+                    Some((files, ts))
+                }
+            })
+            .collect();
```

Comment:
Maybe you can use retain without sorting the vec. Also consider truncating the vec by keep_count after retain.